module/zfs/zfs_vnops.c

```diff
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+	 */
+	if ((dstzp->z_pflags & ZFS_IMMUTABLE)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EPERM));
+	}
+
+	/*
+	 * XXX: There should be no need to sync the source file, as we return
+	 *      an error if somebody is trying to clone a block that was
+	 *      modified in the same transaction group.
+	 */
+#ifdef FRSYNC
+	/*
+	 * If we're in FRSYNC mode, sync out this znode before reading it.
+	 * Only do this for non-snapshots.
+	 *
+	 * Some platforms do not support FRSYNC and instead map it
+	 * to O_SYNC, which results in unnecessary calls to zil_commit. We
+	 * only honor FRSYNC requests on platforms which support it.
+	 */
+	frsync = !!(srcioflag & FRSYNC);
+#endif
+	if (srczfsvfs->z_log &&
+	    (frsync || srczfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)) {
+		zil_commit(srczfsvfs->z_log, srczp->z_id);
+	}
+
+	/*
+	 * No overlapping if we are cloning within the same file.
+	 */
+	if (srczp == dstzp) {
+		if ((srcoffset >= dstoffset &&
+		    srcoffset < dstoffset + length) ||
+		    (dstoffset >= srcoffset &&
+		    dstoffset < srcoffset + length)) {
+			zfs_exit_two(srczfsvfs, dstzfsvfs);
+			return (SET_ERROR(EINVAL));
+		}
+	}
+
+	/*
+	 * Maintain predictable lock order.
+	 */
+	if (srczp < dstzp || (srczp == dstzp && srcoffset < dstoffset)) {
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+	} else {
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+	}
+
+	srcblksz = srczp->z_blksz;
+
+	/*
+	 * Offsets and length must be at block boundries.
+	 */
+	if ((srcoffset % srcblksz) != 0 || (dstoffset % srczp->z_blksz) != 0) {
+		zfs_rangelock_exit(srclr);
+		zfs_rangelock_exit(dstlr);
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EINVAL));
+	}
+	/*
+	 * Length may not be multipe of blksz only at the end of the file.
+	 */
+	if ((length % srcblksz) != 0 && length != srczp->z_size - srcoffset) {
+		zfs_rangelock_exit(srclr);
+		zfs_rangelock_exit(dstlr);
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EINVAL));
+	}
+
+	error = zn_rlimit_fsize(dstoffset + length);
+	if (error != 0) {
+		zfs_rangelock_exit(srclr);
+		zfs_rangelock_exit(dstlr);
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (error);
+	}
+
+	if (srcoffset >= MAXOFFSET_T || dstoffset >= MAXOFFSET_T) {
+		zfs_rangelock_exit(srclr);
+		zfs_rangelock_exit(dstlr);
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EFBIG));
+	}
+
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_MTIME(dstzfsvfs), NULL,
+	    &mtime, 16);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_CTIME(dstzfsvfs), NULL,
+	    &ctime, 16);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_SIZE(dstzfsvfs), NULL,
+	    &dstzp->z_size, 8);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_FLAGS(dstzfsvfs), NULL,
+	    &dstzp->z_pflags, 8);
```

Comment:
Do we ever change z_pflags in this function?

Reply:
No. Removed from here and some other places.